src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -0,0 +1,435 @@
+            action, streamName, count, batches, version)
+
+    let logLastEventRead streamName t events version (log: ILogger) =
+        let bytes = resolvedEventBytes events
+        let count = 1
+        let reqMetric : Log.Measurement = { stream = streamName; interval = t; bytes = bytes; count = count}
+        let evt = Log.Metric.ReadLast reqMetric
+        let batches = 1
+        (log |> Log.prop "bytes" bytes |> Log.event evt).Information(
+            "Mdb{action:l} stream={stream} count={count}/{batches} version={version}",
+            "ReadL", streamName, count, batches, version)
+
+    let loadLastEvent (log : ILogger) retryPolicy (conn : MessageDbClient) streamName
+        : Async<int64 * ITimelineEvent<EventBody> array> = async {
+        let! ct = Async.CancellationToken
+        let read _ = readLastEventAsync conn streamName ct |> Async.AwaitTaskCorrect
+
+        let! t, page = Log.withLoggedRetries retryPolicy "readAttempt" read log |> Stopwatch.Time
+
+        log |> logLastEventRead streamName t page.Messages page.LastVersion
+        return page.LastVersion, page.Messages }
+
+    let loadForwardsFrom (log : ILogger) retryPolicy conn batchSize maxPermittedBatchReads streamName startPosition
+        : Async<int64 * ITimelineEvent<EventBody> array> = async {
+        let mergeBatches (batches : AsyncSeq<int64 * ITimelineEvent<EventBody> array>) = async {
+            let mutable versionFromStream = -1L
+            let! (events : ITimelineEvent<EventBody> array) =
+                batches
+                |> AsyncSeq.map (fun (reportedVersion, events)-> versionFromStream <- max reportedVersion versionFromStream; events)
+                |> AsyncSeq.concatSeq
+                |> AsyncSeq.toArrayAsync
+            let version = versionFromStream
+            return version, events }
+        let call pos = loggedReadSlice conn streamName batchSize pos
+        let retryingLoggingReadSlice pos = Log.withLoggedRetries retryPolicy "readAttempt" (call pos)
+        let log = log |> Log.prop "batchSize" batchSize |> Log.prop "stream" streamName
+        let batches : AsyncSeq<int64 * ITimelineEvent<EventBody> array> = readBatches log retryingLoggingReadSlice maxPermittedBatchReads startPosition
+        let! t, (version, events) = mergeBatches batches |> Stopwatch.Time
+        log |> logBatchRead streamName t events batchSize version
+        return version, events }
+
+module private Token =
+    let create streamVersion : StreamToken =
+        { value = box streamVersion
+          version = streamVersion
+          streamBytes = -1 }
+    let supersedes struct (current, x) =
+            x.version > current.version
+
+type MessageDbConnection(readConnection, [<O; D(null)>]?writeConnection, [<O; D(null)>]?readRetryPolicy, [<O; D(null)>]?writeRetryPolicy) =
+    member _.ReadConnection = readConnection
+    member _.ReadRetryPolicy = readRetryPolicy
+    member _.WriteConnection = defaultArg writeConnection readConnection
```

Comment:
nit: this might as well be a `val` (dont think it matters for the others, but they could also be)

Reply:
I don't think I've ever used a `val` ever 😂. I'll do it and make a note to look up wth it is

![image](https://user-images.githubusercontent.com/3721521/201230868-c85b92a4-0aaa-465d-bfc8-0a16112f014e.png)

Reply:
TL;DR its a property with a backing field thats calculated once/immediately (and e.g. triggers exns immediately), whereas the normal `member` computes every time
sharplab.io is the best way to visually see the impact. In cases where one is just exposing e.g. a primary ctor arg, the IL is identical
@enricosada was the main man that sold it to me - it can clarify a lot of code in argument processing (see the `type Arguments` in dotnet-templates)
In general I prefer it to make "direct exposing of a value that won't change" stick out when scanning the code